# Homework-3 

## Use the Titanic Dataset to Practice Data Wrangling Tasks
### Part 1: Import libaries , load the data, inspect the data, then get the name of the variables and create a small markdown table  with variable names and types (e.g numerical or categorical ordinal, categorical nominal etc.)

### Part 2: 
    - Impute missing values for "Embarked" with the mode (most frequent value)
    - Impute missing value for Age using two following strategies
        - A) Simple mean imputation (fill with the overal mean age)
        - B) Group-wise median imputation: for each combination of Sex and Pclass compute the median age. Impute missing age using the median age of passenger's sex and Pclass group

### Part 3: Create a FamilySize variable (FamilySize = SibSp + Parch +1) and compute:
    - Minimum, Maximum, and mean FamilySize
    - How many passengers were traveling alone
    
### Part 4:  Create a filtered dataframe of all passengers:
    - Who are female
    - Who are male 
    - Who are under 16 years old   
    - Compute the survival rate for the above filtered data
    - Compute survival rate by Sex and Pclass (use groupby and mean) and present the result in a markedown table (rows: sex, columns: Pclass) or using print formating.
    - Interpret your result in 2-3 sentences
     
    
### Part 5: Compute the average fare for each Pclass (Fare values are per tickets not per person (there coud be more than 1 person on a ticket) so make sure to account for that. 

In [9]:
#Part 1
import pandas as pd
import numpy as np

df = pd.read_csv('titanic.csv')

display(df.head())
print(df.info())

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB
None


### Part 1
<table> 
    <tr>
        <th>Variable</th>
        <th>Description</th>
        <th>Data Type</th>
    </tr>
    <tr>
    <td>Survived</td>
    <td>Survival (0 = No, 1 = Yes)</td>
    <td>Categorical Nominal (int)</td>
    </tr>
    <tr>
        <td>Pclass</td>
        <td>Ticket Class (1 = 1st, 2 = 2nd, 3 = 3rd)</td>
        <td>Categorical Ordinal (int)</td>
    </tr>
    <tr>
        <td>Sex</td>
        <td>Sex</td>
        <td>Categorical Nominal (object)</td>
    </tr>
    <tr>
        <td>Age</td>
        <td>Age in years</td>
        <td>Numerical Continuous (float)</td>
    </tr>
    <tr>
        <td>SibSp</td>
        <td># of siblings/spouses aboard</td>
        <td>Numerical Discrete (int)</td>
    </tr>
    <tr>
        <td>Parch</td>
        <td># of parents/children aboard</td>
        <td>Numerical Discrete (int)</td>
    </tr>
    <tr>
        <td>Ticket</td>
        <td>Ticket Number</td>
        <td>Categorical Nominal (object)</td>
    </tr>
    <tr>
        <td>Fare</td>
        <td>Passenger Fare</td>
        <td>Numerical Continuous</td>
    </tr>
    <tr>
        <td>Cabin</td>
        <td>Cabin Number</td>
        <td>Categorical Nominal (object)</td>
    </tr>
    <tr>
        <td>Embarked</td>
        <td>Port of Embarkment (C, Q, S)</td>
        <td>Categorical Nominal (object)</td>
    </tr>
</table>

In [23]:
#Part 2

#fill embarked with the mode
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

#strategy 1
df['mean_age_filled'] = df['Age'].fillna(df['Age'].mean())

#strategy 2
df['median_age_filled'] = df.groupby(['Sex', 'Pclass'])['Age'].transform(lambda x: x.fillna(x.median()))

In [24]:
#Part 3

df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

print(f"Minimum FamilySize: {df['FamilySize'].min()}")
print(f"Maximum FamilySize: {df['FamilySize'].max()}")
print(f"Mean FamilySize: {df['FamilySize'].mean():.2f}")

count_alone = (df['FamilySize'] == 1).sum()
print(f"Passengers traveling alone: {count_alone}")

Minimum FamilySize: 1
Maximum FamilySize: 11
Mean FamilySize: 1.90
Passengers traveling alone: 537


In [25]:
#Part 4

df_female = df[df['Sex'] == 'female']
df_male = df[df['Sex'] == 'male']
df_under_16 = df[df['Age'] < 16]

print(f"Survival Rate (female): {df_female['Survived'].mean()}")
print(f"Survival Rate (male): {df_male['Survived'].mean()}")
print(f"Survival Rate (under 16): {df_under_16['Survived'].mean()}")

grouped = df.groupby(['Sex', 'Pclass'])['Survived'].mean()
print("\n Grouped Survival Rates ")
print(grouped)

Survival Rate (female): 0.7420382165605095
Survival Rate (male): 0.18890814558058924
Survival Rate (under 16): 0.5903614457831325

 Grouped Survival Rates 
Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64


### Part 4

<h2>Survival Rate by Sex and Pclass</h2>

<table>
    <tr>
        <th>Sex</th>
        <th>Class 1</th>
        <th>Class 2</th>
        <th>Class 3</th>
    </tr>
    <tr>
        <td>Female</td>
        <td>0.968085</td>
        <td>0.921053</td>
        <td>0.500000</td>
    </tr>
    <tr>
        <td>Male</td>
        <td>0.368852</td>
        <td>0.157407</td>
        <td>0.135447</td>
    </tr>
</table>

<p>Interpretation: Females survived at a much higher rate than males across all ticket classes. Plus, passengers in first class had a significantly higher survival rate than those in second and third class, most likely due to gender and differing social/economic statuses influencing the emergency protocols, affecting survival rates between classes and gender.</p>

In [26]:
#Part 5

count_tickets_once = df.drop_duplicates(subset=['Ticket'])

average_fare_by_pclass = count_tickets_once.groupby('Pclass')['Fare'].mean()

print("Average Fare for each Pclass (counting shared tickets too):")
print("Pclass 1: $", round(average_fare_by_pclass[1], 2))
print("Pclass 2: $", round(average_fare_by_pclass[2], 2))
print("Pclass 3: $", round(average_fare_by_pclass[3], 2))

Average Fare for each Pclass (counting shared tickets too):
Pclass 1: $ 64.14
Pclass 2: $ 17.51
Pclass 3: $ 10.08
